### ISSUE TITLE
Arguments set to defaults of other arguments

### ISSUE THREAD

**Issue by kaaaahhhhnnnn** [babel/babel#1896]
Okay, I have no idea what the correct behavior is here, or if it's even allowed, but:

```js
(a=b, b)=> {};
(a=b, b=1)=> {};
```

The first case works, oddly enough.  In the 2nd case, I would _expect_ `a==b==1`, but it results in `a==undefined, b==1`

**Comment by eventualbuddha** [comment]
Can you give a more complete example from the [REPL](babeljs.io/repl/)?

**Comment by kaaaahhhhnnnn** [comment]
Paste that into the repl :)

The 2nd function yields:

```js
  var a = arguments[0] === undefined ? b : arguments[0];
  var b = arguments[1] === undefined ? 1 : arguments[1];
```

meaning that `a=b` gets set before `b=1`.  I've no idea if that's correct or not according to spec, but it seems like it shouldn't be

**Comment by Mark-Simulacrum** [comment]
@kaaaahhhhnnnn You can link to code on the Babel REPL: [@kaaaahhhhnnnn's example](http://babeljs.io/repl/#?code=(a%3Db%2C%20b)%3D%3E%20%7B%7D%3B%0A(c%2C%20a%3Db%2C%20b%3D1)%3D%3E%20%7B%7D%3B)

**Comment by kaaaahhhhnnnn** [comment]
@Mark-Simulacrum -- cool, never noticed that :)

**Comment by eventualbuddha** [comment]
I think default argument values should only be allowed to refer to bindings before their own binding definition, otherwise you could end up with a mess of cycles with no way for the compiler to resolve it:

```js
(a=b, b=a) => {};
```

And what I wanted from a REPL link was a "run it like this and I expect to see this output but I got this", something like:

```js
// expected "a = 1 b = 1", got "a = undefined b = 1"
((a=b, b=1) => { console.log('a = ' + a + ' b = ' + b); })();
```

As I said, I think this construct should not be allowed.

**Comment by kaaaahhhhnnnn** [comment]
fair enough, wasn't considering circular references.  Thanks for your time :)

**Comment by Mark-Simulacrum** [comment]
I believe the purpose of the construct is to allow the following pattern, where both variables are defined and valid. (No consultation of spec, don't know if this works, but I would assume that it does.)
```js
let default = "default value";
function foo(a = default, default = true) {
  // a = "default value"
  // b = true
}
```

**Comment by kaaaahhhhnnnn** [comment]
yeah, it works going forwards, just not backwards.  I was thinking it should work in circumstances like from `c->a->b`, but the `c->a->b->c` case kinda breaks that :)

**Comment by kaaaahhhhnnnn** [comment]
Although yikes, babel and closure differ here...

@Mark-Simulacrum -- With your code, babel uses the external variable, closure uses the internal one.  Any idea which is actually correct?

**Comment by kaaaahhhhnnnn** [comment]
okay, I'm pretty sure _something_ is off here:

```
let b = "default value";
foo(a=b, b)

**Comment by Mark-Simulacrum** [comment]
I would assume that both are correct. The default arguments create a "scope of themselves" if I remember correctly, so `default` in `a = default` refers to the `let default` declaration in the parent scope. `default = true` is defining an argument (for the function's scope).

As far as I know, even this code is valid:
```js
let a = "first value of a";
function foo(a = a) {
  // a = arguments[0] === undefined ? "first value of a" : arguments[0];
}
````

The code that you are posting *seems* valid to me: a function can have arguments that shadow the definition(s) of parent scopes'.

**Comment by kaaaahhhhnnnn** [comment]
```js
(a=b,b=1) => {console.log(a,b)}(undefined, 43);
```

expected `43, 43` got `undefined, 43`.  FWIW closure yields `43, 43`

**Comment by Mark-Simulacrum** [comment]
When you define `a = b` for the first argument; this examines **just** the value of `arguments[0]`. Babel (and the spec, as far as I know), will not attempt to determine if the right hand side of the default argument refers to an argument farther down the chain. The arguments are evaluated at that point in the code.

I expect `((a=b,b=1) => console.log(a,b))(undefined, 43);` to yield `undefined, 43`;
when `a = b` is examined, Babel sees that as "is `arguments[0]` defined? If yes, then `arguments[0]` otherwise `b`". No attempt is made, or should be made (I believe), to determine if `b` evaluates to an argument further in the list of arguments.

**Comment by kaaaahhhhnnnn** [comment]
If that's the case, I would expect:

```js
let b=42;
((a=b,b=1) => console.log(a,b))(undefined, 43);
```

to yield `42, 43`, but it still yields `undefined, 43`

**Comment by Mark-Simulacrum** [comment]
The reason appears to be that I was wrong (or so it seems) about how default arguments are evaluated. If "high compliancy" mode is enabled in the REPL for [this](http://babeljs.io/repl/#?spec=true&code=let%20b%3D42%3B%0Afunction%20foo(a%20%3D%20b%2C%20b%20%3D%201)%20%7B%0A%20%20console.log(a%2Cb)%0A%7D%20%0A%0Afoo(undefined%2C%2043)%3B) code, we see that the reason that we see `undefined, 43` is because `var b` is being hoisted to the top of the function scope, which is creating this seemingly inconsistent behavior.

In simpler terms, arguments in the list are hoisted to the top of the list: `function (a, b) {}` is really:
```js
function () {
  var a;
  var b;
  a = arguments[0];
  b = arguments[1];
}
```

However, I am no longer certain that Babel correctly implements this feature (and have been unable to find default arguments in the spec in order to attempt to check). I still think that it is correctly implemented, but can't be 100% certain. @sebmck should be able to clear the question up, though.

**Comment by loganfsmyth** [comment]
My understanding is that arguments are processed left to right, and that arguments are declared one at a time, so they have temporal dead zone rules just like `let`. I would describe the expected behavior of

    (a=b, b=1)=> {};

as

```
function fn(){
    let a = arguments[0] === undefined ? b : arguments[0];
    let b = arguments[1] === undefined ? 1 : arguments[1];
}
```

which means that `b` is accessed before it has been declared, leading to a `let` temporal dead zone violation. If you enable "high compliancy" in the repo, or the transform `es6.spec.blockScoping` in your configuration.

I'd definitely recommend enabling the spec transformers when doing local development at least, though they slow things down so probably not in production builds.

**Comment by sebmck** [comment]
```javascript
(a=b, b)=> {};
(a=b, b=1)=> {};
```

is a temporal dead zone error since you're referencing a variable before it's initialised. This is very hard to replicate with ES5 semantics. You can enable the optional `es6.spec.blockScoping` transformer which will produce verbose code but catch this case by delegating the checking to the runtime.
